packages/core/src/internal/io.js

```diff
@@ -209,7 +210,7 @@ export function setContext(props) {
 
 export const delay = call.bind(null, delayUtil)
 
-const createAsEffectType = type => effect => effect && effect[IO] && effect[type]
+const createAsEffectType = type => effect => effect && effect[IO] && effect.type === type && effect.payload
 
 export const asEffect = {
```

Comment:
Maybe we can remove `asEffect`, and place the logic in runEffect. So in `proc.js` it would look like:
```javascript
function runEffect(effect, effectId, currCb) {
  if(effect[IO] === true) {
    const { type, payload } = effect
    if (type === 'TAKE') {
      runTakeEffect(payload, currCb)
    } else if (type === 'PUT') {
      runPutEffect(payload, currCb)
    } // ......
  }
}
```
`asEffect` is exported by `redux-saga/utils`.  Maybe it is not a good idea to refactor `asEffect`. How do you think about it?

Reply:
I don't know how `asEffect` was used outside of the core, let's just ditch it for now & wait to see if it breaks anyone's use case. We are in v1-beta period right now, so I'm not concerned with making some breaking changes.

I like refactoring `runEffect` in the way you have proposed here.

Reply:
`asEffect` seems like a inspection/test utils. (#1260)
I will continue refactoring `runEffect`. If we refactor the effect structure, it would be more easier for users to do inspections/tests.